src/tools/fuzzing/fuzzing.cpp

```diff
@@ -693,9 +693,77 @@ Expression* TranslateToFuzzReader::makeHangLimitCheck() {
 }
 
 Expression* TranslateToFuzzReader::makeLogging() {
-  auto type = getLoggableType();
-  return builder.makeCall(
-    std::string("log-") + type.toString(), {make(type)}, Type::none);
+  auto makeLoggingCall = [&](Expression* value) {
+    assert(isLoggableType(value->type));
+    return builder.makeCall(
+      std::string("log-") + value->type.toString(), {value}, Type::none);
+  };
+
+  // We may choose to log a local, if there are any.
+  auto* func = funcContext->func;
+  auto numLocals = func->getNumLocals();
+
+  auto choice = upTo(4);
+  if (choice < 2 || numLocals == 0) {
+    // 50% of the time, pick a loggable type and make something of that type
+    // to log. Also do so when there are no locals to log.
+    return makeLoggingCall(make(getLoggableType()));
+  }
+
+  // 50% of the time, pick a local and log it, either shallowly or deeply.
+  auto index = upTo(numLocals);
+  auto type = func->getLocalType(index);
+  if (isLoggableType(type)) {
+    // We can log this directly. Get it and log that.
+    return makeLoggingCall(builder.makeLocalGet(index, type));
+  }
+
+  if (type.isRef()) {
+    // This is a reference, which cannot be directly logged. We can at least
+    // "shallowly" log it by seeing if it is null (if it is nullable - if not,
+    // then there is no point to such a check).
+    auto* get = builder.makeLocalGet(index, type);
+    auto* isNullCheck =
+      type.isNullable() ? builder.makeRefIsNull(get) : nullptr;
+    if (choice & 1) {
+      // Try to also "deeply" log it, by reading a value from it, if we can.
+      auto heapType = type.getHeapType();
+      if (heapType.isStruct()) {
+        auto& fields = heapType.getStruct().fields;
+        if (!fields.empty()) {
+          auto fieldIndex = upTo(fields.size());
+          auto fieldType = fields[fieldIndex].type;
```

Comment:
Do you think it's worth linearly searching from this random starting index looking for the first loggable field? We could also do a breadth-first search on the type graph looking for a loggable type.